### DISCUSSION TITLE
How to trigger an automation (action) with a button.

### DISCUSSION THREAD

**Discussion by Zaleo80** [Blackymas/NSPanel_HA_Blueprint#810] [General]
I would like to trigger an automation when a button is clicked.
Maybe I am misunderstanding it, but it seems like the automation is enabled/disabled instead of triggered.

**Comment by deejaybeam** [comment]
If you want to trigger an automation, this has to be done with the trigger-section in the automation itself.
I think you want to execute someting when a button is pressed -> than you have to use a script for that.
The functionality to enable/disable an automation is e.g. to disable heating-automation in summer...

**Comment by hkr086** [comment]
If it's a virtual button you are looking for then use an input Boolean which would be used in your automation as a trigger

**Reply by ZeoSmeijsters** [reply]
> If it's a virtual button you are looking for then use an input Boolean which would be used in your automation as a trigger

Yes, that is the way I will solve it.

However I find it strange that for holding the fysical buttons I can add any action but for screen buttons (and normal pressing a fysical button) we can't.

**Reply by Doughboy68** [reply]
Yes would be nice to have a page with custom action buttons just like the hard buttons work, but the blueprint interface would get more complex.  A helper is the better way to do it. While a Button Helper would also work, I can see using a Toggle (Input Boolean) and have automation turn it off so can see visual update on panel, then  add a few seconds of wait in the automation and then turn the Toggle back off... would look nice on panel for it to "light up" and give visual feedback for a few seconds when pressed. I am thinking now of doing some helpers/automations for a fan speed and dedicate a page to fan controls... have one automation with four input booleans, and use trigger options to adjust them appropriately to set fan lo,med,hi, off.

**Comment by Zaleo80** [comment]
From my understanding the goal is to make it a NoCode solution for the NSPanel, and you are doing a great job.

However setting any action with a button instead of only one entity with a (autocomplete) input box would not divert from this goal. Besides if someone does not know how to set an action they can't even setup an automation in HA. Don't see this persona flash the NSPanel in the first place.

Therefore I do not see the argument that it would get to complex. You would only replace the input box with a button.

Secondly I think this is more in line with the HA UI and therefore more friendly.

Only thing that might be tricky is the conversation during an update.

Just a thought.

**Comment by edwardtfn** [comment]
So, this topic is coming back again with v4.1 and I'd like your help to find a better way forward. 😃 

I've changed the behavior of automation in a button after #1249 from @rudgej, however @samjw-nz came back to this on #1291 and looks like we have the exactly same conflict from this discussion (but also from #320 and #331).

My first thought was "a button executes an action", so it was clear to me that the automation should run, but now reading all this material (ok, I should have done before changing the code) I also agree that "an automation without a trigger is a script".

I'm not gonna try to find a consensus on those, but I wanna find a way to implement this. I'm fine about having a global input on the Blueprint for users define how automations on buttons should behave (`trigger` vs `enable/disable`), but before jumping on that, I'd like to hear your thoughts and I'm totally open for new ideas.

Please speak out. 😃

**Reply by Doughboy68** [reply]
Automations' actions are triggered by events in Home Assistant. Triggering by a button press from NSPanel directly via blueprint does not make sense to me. Automations can have multiple triggers in Home Assistant so how is it supposed to know what action to run when it is simply run?  Most of my automations do nothing when I run them because there is no trigger.

If one is running simple automations with no real triggers or conditions, then they should be using  a script instead, which can run by a button in blueprint. If also want to automate it, simply make an automation with a trigger like a time or something that has an action which runs same script.

I have a few buttons on NSPanel which run scripts that set RGB color scenes on multiple lights.

Therefore, I recommend putting reverting back so a blueprint button toggles automation on/off. Still not sure why one would want to do this anyways to enable/disable an automation by a button? Could instead use an input boolean helper and integrate that logic into a condition inside an automation... that way can also make buttons in HA dashboards to enable/disable functions. Example... I have an automation that monitors a lift pump, and I have an input boolean which enables/disables part of that automation to send a notification on my phone when the pump runs for troubleshooting.

**Comment by samjw-nz** [comment]
Appreciate circling back to this. While I agree with the concept that 'a button executes an action' in theory, in most cases for the NSPanel I imagine the hardware buttons replace a toggle switch in practice (a light switch, by toggling the state of a relay). 

While it seems possible but not entirely practical to trigger a script to toggle an automation, my opinion is that should be secondary to the expected default of switching something on/off (i.e a light bulb, or an automation).
One way or another, it's likely a workaround will be needed for either side of the argument. I just think the hardware buttons should default to operating as switches, as that's what most people would be replacing with the NSPanel anyway.

Alternatively (or additionally?) having direct control of the status lights would allow more flexibility on how they're used. I use one hardware button to trigger a full-bright scene for my Philips Hue lights, but because I can't set the status light to be on/off I cannot look at the display to tell me the on/off status of my lights. 
A pointless example if the lights are in the same room, but if it controlled lights in another room how do I know if they're on or off?
Honestly this one isn't a huge deal for me, but throwing it out there as a thought anyway.

**Reply by Doughboy68** [reply]
Off topic, but I use a light group and then an add-on called "adaptive lighting" that allows me to do this. I basically have a preset color temp and brightness for when the light turns on from powered off. I do not use any of the adaptive settings and make it so it just keeps the same setting. This way I can mess around with changing colours and brightness etc. and when I turn light off and on it has a default normal color.

**Reply by deejaybeam** [reply]
@samjw-nz you don´t need to use a script to trigger an automation - just use the script to do what the automation itself does.
everything in the action part of an automation can be copy-pasted into a script, but without the trigger-part. Thats the purpose of a script.
Anyways a scene, a script and an automation are always "stateless" you can trigger the execution, but you will never see something like on/off like a light does. That means you never will see if a scene or script is "current active".
For me I also have the use-case to use automation enable/disable:
With this I enable/disable the automation to control the room-climate (like switching between summer/winter heatinig-mode)

**Reply by edwardtfn** [reply]
> you don´t need to use a script to trigger an automation - just use the script to do what the automation itself does.
everything in the action part of an automation can be copy-pasted into a script, but without the trigger-part. Thats the purpose of a script.

And you don't have to create a redundant script, as you can move the automation entirely to the script and in the automation itself having just the triggers and conditions, the the action woukd be executing that script, so only one point for maintenance.

**Reply by Linwood-F** [reply]
I am doing something similar, each button activates a script which sets a specific mode of landscape lighting.  If you hit it again, it turns the lights off (this is handled and persisted inside the script).  Right now I indicate those with chips at the top, but it would sure be nice to have a link between the bar/label over the button instead of a chip, more intuitive.  In my case the work is all done in HA, I have an entity already used in the chip to light it up (though changing color would be nice).

**Comment by rudgej** [comment]
I seem late to the discussion so apologies. I was also the person who questioned the current behaviour as I expected the automation to run. I can see a common use case for turning an automation on  and off.  Not no uses case - but a common one- the heating on / off on change of season will happen once a year.

My home - like other I assume - are run by automations. Evening on, Evening Off, Curtains closed and so on. These automations typically call multiple scenes so that each room goes to " Evening " or "Off" etc.

I am OK at Home Assistant because I follow the instructions rather than being particularly gifted. I don't know how a button on the panel can call a script and I don't know how to write one! Assuming I did then the script would simply have to re create an automation that already exists-  giving 2 point of update each time something is modified. 

I started a discussion to which only Edward replied as to " how do people use their panels" as I felt looking at the feature requests many are niche to a user, and if there were a better understanding of how these are used. practice ( lighting scene controllers in my view) then the features might be more aligned with the majority. 

On this issue presumably there could be a toggle to select if the automation runs or is turned on /off - but running the automation in my view would seem by far the most popular use case. Other of course may have a differing view.

If you revert it then can someone please tell me how I make a button run an automation via another method!? 

Jeremy

**Comment by Doughboy68** [comment]
A script is similar to an automation but does not have a trigger. You can put your code in a script which turns lights on etc and call it from an automation. You can assign script  to button in nspanel to run also. There is no two points to update - you are just moving the action code into a script.  

It doesn’t make sense to run automations without triggers. Many of my automations have several triggers and do things depending on what triggered therefore assigning to a button doesn’t make sense. It doesn’t affect me either way since I would never assign an automation to a button, however I would prefer to see Nspanel adhere to common standards and practices without implementing workarounds for, in this case, something which seems like should be using script. 

https://youtu.be/kWp-XQRcmmM?si=-bXmFJf7X7vekthS

**Comment by edwardtfn** [comment]
I've added a selector to switch between **Toggle** and **Trigger** for button automations. This feature is now in the `dev` branch and will be included in our next release.
